From ed2eeacc0a8cf8bdc69be20e6f1f590a1e76c203 Mon Sep 17 00:00:00 2001
From: Swati Singh <swasing@qti.qualcomm.com>
Date: Wed, 18 Mar 2026 11:28:25 +0530
Subject: [PATCH] [linux] Enable Port Isolation in Linux.

Port isolation is enabled for downstream PON traffic.
The skb_ext metadata includes the sub_br_id,
which is referenced in the br_flood function to
restrict packet flooding to interfaces within
the same bridge domain.

Change-Id: Id0fbbf02ba666380c12adf3fc911547768aab231
Signed-off-by: Swati Singh <swasing@qti.qualcomm.com>
---
 include/linux/skbuff.h  | 12 ++++++++++++
 net/bridge/br_forward.c | 24 ++++++++++++++++++------
 net/bridge/br_input.c   | 24 +++++++++++++++++++-----
 net/core/skbuff.c       |  3 +++
 4 files changed, 52 insertions(+), 11 deletions(-)

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index 842d4b458bf7..47a69b7fd492 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -352,6 +352,15 @@ struct sk_buff_head {
 	spinlock_t	lock;
 };
 
+#ifdef CONFIG_IPQ_PON
+struct gem_skb_ext {
+	u32 sub_br_id;
+	u8 gem_id;
+	u8 flood;
+	u8 __pad[2];
+};
+#endif
+
 struct sk_buff;
 
 #ifndef CONFIG_MAX_SKB_FRAGS
@@ -4801,6 +4810,9 @@ enum skb_ext_id {
 #endif
 #if IS_ENABLED(CONFIG_MCTP_FLOWS)
 	SKB_EXT_MCTP,
+#endif
+#ifdef CONFIG_IPQ_PON
+	SKB_EXT_GEM,
 #endif
 	SKB_EXT_NUM, /* must be last */
 };
diff --git a/net/bridge/br_forward.c b/net/bridge/br_forward.c
index 4acaa2fd3915..fbddccaed68f 100644
--- a/net/bridge/br_forward.c
+++ b/net/bridge/br_forward.c
@@ -213,7 +213,10 @@ void br_flood(struct net_bridge *br, struct sk_buff *skb,
 	struct net_bridge_port *srcp = br_port_get_rcu(skb->dev);
 	struct net_bridge_port *prev = NULL;
 	struct net_bridge_port *p;
-
+	u16 ref_sub_br_id;
+#ifdef CONFIG_IPQ_PON
+	struct gem_skb_ext *gem_ext = skb_ext_find(skb, SKB_EXT_GEM);
+#endif
 	br_tc_skb_miss_set(skb, pkt_type != BR_PKT_BROADCAST);
 
 	list_for_each_entry_rcu(p, &br->port_list, list) {
@@ -248,11 +251,20 @@ void br_flood(struct net_bridge *br, struct sk_buff *skb,
 			continue;
 
 		/* Do not flood to non-upstream port and to ports in different sub bridge */
-		if (srcp &&
-			!((p->flags & BR_UPSTREAM_PORT) || (srcp->flags & BR_UPSTREAM_PORT))
-			&& (p->sub_br_id != srcp->sub_br_id))
-			continue;
-
+		if (srcp && !((p->flags & BR_UPSTREAM_PORT) || (srcp->flags & BR_UPSTREAM_PORT))) {
+			/*
+			 * Flood only if dest port is in the same sub-bridge as:
+			 *  - Source port's sub-bridge id, OR
+			 *  - GEM OMCI sub-bridge id when GEM flood is set.
+			 */
+			ref_sub_br_id = srcp->sub_br_id;
+#ifdef CONFIG_IPQ_PON
+			if (gem_ext && gem_ext->flood)
+				ref_sub_br_id = gem_ext->sub_br_id;
+#endif
+			if (p->sub_br_id != ref_sub_br_id)
+				continue;
+		}
 
 		prev = maybe_deliver(prev, p, skb, local_orig);
 		if (IS_ERR(prev))
diff --git a/net/bridge/br_input.c b/net/bridge/br_input.c
index 7aab36d438a3..67be3a178c85 100644
--- a/net/bridge/br_input.c
+++ b/net/bridge/br_input.c
@@ -99,6 +99,10 @@ int br_handle_frame_finish(struct net *net, struct sock *sk, struct sk_buff *skb
 	br_get_dst_hook_t *get_dst_hook = rcu_dereference(br_get_dst_hook);
 	u16 vid = 0;
 	u8 state;
+	u16 ref_sub_br_id;
+#ifdef CONFIG_IPQ_PON
+	struct gem_skb_ext *gem_ext = skb_ext_find(skb, SKB_EXT_GEM);
+#endif
 
 	if (!p)
 		goto drop;
@@ -232,12 +236,22 @@ int br_handle_frame_finish(struct net *net, struct sock *sk, struct sk_buff *skb
 		} else {
 			dst = br_fdb_find_rcu(br, eth_hdr(skb)->h_dest, vid);
 
-			/* Restrict forwarding to/from upstream port & within same sub bridge */
+			/* Do not flood to non-upstream port and to ports in different sub bridge */
 			if (dst && dst->dst &&
-				!((p->flags & BR_UPSTREAM_PORT) || (dst->dst->flags & BR_UPSTREAM_PORT))
-				&& (p->sub_br_id != dst->dst->sub_br_id))
-				dst = NULL;
-
+					!((p->flags & BR_UPSTREAM_PORT) || (dst->dst->flags & BR_UPSTREAM_PORT))) {
+				/* Flood only if source sub bridge id same as
+				 *  - Dest port's sub-bridge id, OR
+				 *  - GEM OMCI sub-bridge id when GEM flood is set.
+				 */
+				ref_sub_br_id = p->sub_br_id;
+#ifdef CONFIG_IPQ_PON
+				if (gem_ext && unlikely(gem_ext->flood))
+					ref_sub_br_id = gem_ext->sub_br_id;
+#endif
+				if(skb->pkt_type != PACKET_HOST)
+					if (dst->dst->sub_br_id != ref_sub_br_id)
+						dst = NULL;
+			}
 		}
 		break;
 	default:
diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index d36f2ae48eb8..fb59a5365017 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -5175,6 +5175,9 @@ static const u8 skb_ext_type_len[] = {
 #if IS_ENABLED(CONFIG_MCTP_FLOWS)
 	[SKB_EXT_MCTP] = SKB_EXT_CHUNKSIZEOF(struct mctp_flow),
 #endif
+#ifdef CONFIG_IPQ_PON
+	[SKB_EXT_GEM] = SKB_EXT_CHUNKSIZEOF(struct gem_skb_ext),
+#endif
 };
 
 static __always_inline unsigned int skb_ext_total_length(void)
-- 
2.34.1

